Paying to each holder in each group

joshuef opened the pull request:
starting to simplify payments to a per chunk basis

A first step in #638## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 16 Aug 23 14:29 UTC
This pull request includes the following changes:
1. In the `lib.rs` file:
   - Import `PublicKey` from `libp2p::identity` and use it in addition to `Multiaddr` from `libp2p`.
   - Add a new field `reward_main_key` of type `PublicKey` to the struct `Node`.
2. In the `file_apis.rs` file:
   - Import statement added: `use crate::NetworkAddress`.
   - Import statement removed: `use crate::messages::PaymentProof`.
   - Import statement added: `use sn_dbc::DbcId`.
   - Method added: `network_address()`, returns the `NetworkAddress` associated with the chunk.
   - Struct field type changed: `payment` changed from `PaymentProof` to `Vec<DbcId>`.
3. In the `api.rs` file:
   - Added `PublicKey` import from `identity` module.
   - Renamed method `get_store_cost_from_network` to `get_store_costs_from_network` and changed its return type to `Result<Vec<(PublicKey, Token)>>`.
   - Updated error handling and added decoding of `PublicKey` in `get_store_costs_from_network` method.
   - Added a retry loop for `put_record_with_retries` method.
   - Made `get_closest_peers` method public.
4. In the `storage_payments.rs` file:
   - Import `sn_dbc::{Hash, Token}` has been removed.
   - The type of `random_content_addrs` has been changed from `XorName` to `sn_protocol::NetworkAddress::ChunkAddress`.
   - The `eyre::Result` import has been simplified.
   - Various changes related to the variable `invalid_proofs` have been made in the function `storage_payment_chunk_upload_fails`.
5. In the `mod.rs` file in the `sn_protocol/src/messages` directory:
   - In the `use` statement, `PaymentProof` has been replaced with `PaymentTransactions`.
6. In the `data_with_churn.rs` file:
   - The method `pay_for_storage` is modified to use the `network_address()` method instead of the `name()` method on each element of the `chunks` iterator.
7. In the `verify_data_location.rs` file:
   - There is a change in the `store_chunk` function where the `pay_for_storage` method is called with the `network_address()` method instead of the `name()` method on each element of the `chunks` iterator.
8. In the `sn_cli/src/subcommands/files.rs` file:
   - The import statement `sn_transfers::wallet::PaymentProofsMap` has been replaced with `sn_transfers::wallet::PaymentTransactionsMap`.
   - The argument `payment_proofs` in the function `upload_chunks` has been changed from `&PaymentProofsMap` to `&PaymentTransactionsMap`.
9. In the `wallet.rs` file:
   - The type `PaymentProofsMap` has been renamed to `PaymentTransactionsMap`.
   - The `get_payment_proof` method now accepts a `NetworkAddress` parameter instead of `XorName`, and returns an `Option<&Vec<DbcId>>` instead of `Option<&PaymentProof>`.
   - The `add_payment_proofs` method now accepts a `PaymentTransactionsMap` parameter instead of `PaymentProofsMap`.
   - The `local_send_storage_payment` method now accepts a `storage_cost_for_addr` parameter of type `Token` instead of `storage_payment` and `root_hash`.
   - The `wallet.paymet_proofs` field has been renamed to `wallet.payment_transactions`.
10. In the `.github/workflows/memcheck.yml` file:
    - Line 101: The `timeout-minutes` value has been updated from 25 to 40.
11. In the `sn_networking/src/error.rs` file:
    - Added `use libp2p::identity::DecodingError;`.
    - Added variant `Libp2pDecode` to the `Error` enum with a `DecodingError` parameter.
12. In the `sn_protocol/src/messages/cmd.rs` file:
    - The struct `PaymentProof` has been renamed to `PaymentTransactions`.
    - The fields `audit_trail` and `path` of the `PaymentTransactions` struct have been removed.
13. In the `sn_client` module:
    - The `network_store_cost` field has been removed from the `Client` struct.
14. In the `transfer.rs` file:
    - Line 66: The `root_hash` variable is no longer passed as a parameter to the function `create_storage_payment_transfer()`.
    - Line 70: Added a comment to indicate that some code needs to be cleaned up in the future.
    - Line 72: The `root_hash` variable is now created using the `sn_dbc::Hash::hash()` function and the value "nonsense" as input.
15. In the `sn_protocol/src/messages/mod.rs` file:
    - The import statement for `sn_protocol::storage::ChunkAddress` has been removed.

Please review these changes and let me know if you have any questions or need further assistance.
<!-- reviewpad:summarize:end -->

RolandSherwin committed:
fix(node): handling events should wait before connected to the network

joshuef committed:
feat!(protocol): remove chunk merkletree to simplify payment

starting to simplify payments to a per chunk basis

A first step in #638

joshuef committed:
feat!(protocol): get price and pay for each chunk individually

joshuef committed:
feat!(protocol): gets keys with GetStoreCost

joshuef committed:
feat!: pay each chunk holder direct

Sets us up to pay every node that will be storing the data.
We GetStoreCost and pay to their key, and put the chunk with the DBCs

This is not yet validated at the node side

joshuef committed:
feat(node): store data if the majority of CLOSE_GROUP will

joshuef committed:
test(node): reenable payment fail check

joshuef committed:
chore(node): clarify payment errors

joshuef committed:
chore(node): only store paid for data, ignore maj

joshuef committed:
chore(networking): return all GetStoreCost prices and use them

This gives maximal resilience as a place to start.
(This is required for data location verification tests eg)

joshuef committed:
chore(client): error out early for invalid transfers

joshuef committed:
feat: one transfer per data set, mapped dbcs to content addrs

RolandSherwin commented on sn_client/src/api.rs:
```diff
@@ -315,27 +313,28 @@ impl Client {
     }
 
     /// Get a dbc spend from network
+    /// Optionally pass in an expected Dbc if we're verifying a spend from the client
```
Do we still optionally pass in a Dbc?

RolandSherwin commented on sn_client/src/wallet.rs:
```diff
@@ -104,86 +103,103 @@ impl WalletClient {
+        let mut payment_map = BTreeMap::default();
+
+        let mut tasks = JoinSet::new();
+        // we can collate all the payments together into one transfer
```
What do we mean by one transfer here? We do 1 transfer per record right?

joshuef replied:
yeh, one per record is what we have right now. That could be improved upon I think

RolandSherwin commented on sn_client/src/wallet.rs:
```diff
@@ -104,86 +103,103 @@ impl WalletClient {
+            // otherwise all pushes are waiting on all prior verifications
+            self.client.send(transfer.clone(), false).await?;
+
         }
```
Maybe for the follow up PR? I think we can parallelize the `sends` here?

joshuef replied:
hmm, things get complicated here, as they may actually be chained. I thin I'm going to remove this and we have to look at what makes sense to do in parallel here as optimisations down the line.

RolandSherwin commented on sn_networking/src/error.rs:
```diff
@@ -64,6 +68,8 @@ pub enum Error {
 
     #[error("Dial Error")]
     DialError(#[from] DialError),
+    #[error("Libp2p Idendity Decode Error")]
```
```suggestion
    #[error("Libp2p Identity Decode Error")]
```

RolandSherwin commented on sn_networking/src/lib.rs:
```diff
@@ -704,6 +708,13 @@ impl Network {
+                    {
+                        error!("Record check: expected {target_record:?}");
+                        error!("Returned: {returned_record:?}");
+                        return Err(Error::ExpectedNoRecordToExist(returned_record.key.into()));
```
Should we also return error if `target_record.is_none()`? 
Also can we add some docs for this argument?

RolandSherwin commented on sn_protocol/src/error.rs:
```diff
@@ -67,11 +67,11 @@ pub enum Error {
-    #[error("The id of the fee output found in a storage payment proof is invalid: {}", .0.to_hex())]
-    PaymentProofInvalidFeeOutput(Hash),
+    #[error("The id of the fee output found in a storage payment proof is invalid: {0:?}")]
+    PaymentProofInvalidFee(Token),
```
This error is unused and can be removed.

joshuef committed:
test(node): data verification test refactors for readability

joshuef committed:
chore(node): data verification log tweaks

maqi committed:
fix: not check payment for relocated holder

maqi committed:
fix: correct replicated spend validation

RolandSherwin committed:
fix(network): trigger bootstrap until we have enough peers

joshuef commented on sn_node/src/put_validation.rs:
```diff
@@ -230,16 +231,32 @@ impl Node {
             None => {
                 // data is already present
-                return Ok(CmdOk::DataAlreadyPresent);
+                if is_new_put {
```
if it's a new put, can the data already be present? I'm not sure I follow this change

maqi replied:
If it's a new put, it's just follow the current flow.
The put could be fired by mistake or say `second put try to ensure data uploaded`

joshuef committed:
ci: add missing package specification to dbc spend tests

joshuef committed:
ci: increase node mem limit with increased dbc processing

joshuef committed:
fix(protocol): avoid panics

joshuef committed:
chore: increase concurrent fetches for replication data

joshuef committed:
test: parallelise churn data final query

joshuef committed:
chore(networking): ensure we're always driving forward replication if pending

joshuef committed:
chore(client): reduce transferoutputs cloning

joshuef committed:
chore(client): pass around content payments map mut ref

joshuef committed:
chore: logs

joshuef committed:
ci: Temp mem increase.

Holding DBCs causes large mem spikes. For now we allow for this on CI/
We'll refactor this away down the line

maqi committed:
chore: mem_check test update

RolandSherwin committed:
chore(deps): bump tokio to 1.32.0
